src/v/model/ktp.h

```diff
@@ -0,0 +1,347 @@
+     * object k, it is always the case that:
+     *
+     * k == k.to_ntp()
+     *
+     * This object returns by-value, so in general may perform
+     * an allocation for the topic name if it is longer than
+     * the sstring SSO threshold: the "kafka" namespace is
+     * smaller than the SSO threshold and won't allocate.
+     */
+    ntp to_ntp() const { return {kafka_namespace, tp}; };
+
+    /**
+     * @brief Return a topic namespace view corresponding to this object.
+     *
+     * Return a topic_namespace_view with kafka as the namespace and
+     * this topic.
+     *
+     * The view points to inside this object, so its lifetime (or at least
+     * the accesses to the view) must be nested within this object's lifetime.
+     *
+     * Does not allocate as the view simply points to existing objects.
+     */
+    topic_namespace_view as_tn_view() const {
+        return {kafka_namespace, get_topic()};
+    }
+
+    /**
+     * @brief Return a topic partition view corresponding to this object.
+     *
+     * Return a topic_partition_view this objects topic and partition. Note
+     * that a topic_partition_view does not have an implicit "kafka" namespace
+     * so this conversion loses some information.
+     *
+     * The view points inside this object, so its lifetime must be a subset of
+     * this object's lifetime.
+     *
+     * Does not need to allocate as the view points to existing objects.
+     */
+    topic_partition_view as_tp_view() const {
+        return {get_topic(), get_partition()};
+    }
+
+    /**
+     * @brief Returns a reference to the ktp's topic.
+     */
+    const model::topic& get_topic() const { return tp.topic; }
+
+    /**
+     * @brief Returns the ktp's partition id.
+     */
+    model::partition_id get_partition() const { return tp.partition; }
+
+    /**
+     * @brief Returns the ktp's namespace as a string_view.
+     *
+     * Of course, the namespace is always "kafka", but this method
+     * is provided to ease interop with other types which may have
+     * variable namespaces.
+     */
+    constexpr const std::string_view get_ns() const { return kafka_ns_view; }
+
+    /**
+     * @brief Compare this object with another ktp.
+     *
+     * This is more efficient than ntp-only or mixed ntp <-> ktp comparison
+     * since it does not need to compare the namespace at all.
+     */
+    bool operator==(const ktp& other) const { return tp_equals(tp, other.tp); }
+
+    /**
+     * @brief Compare this object with an ntp.
+     *
+     * Although this must compare the namespaces it is still in practice faster
+     * than an ntp <-> ntp comparison, because one side of the namespace
+     * comparison the constexpr string "kafka", which leads to very efficient
+     * code generation (an inlined, fixed-size memcmp) for that comparison.
+     */
+    bool operator==(const ntp& other) const {
+        return tp_equals(other.tp, tp) && kafka_ns_view == other.ns();
+    }
```

Comment:
Can you clarify a bit about what problems this might cause? I expect the order of arguments not to matter here and there are unit tests to check it.

Reply:
Yeah seen those tests and I'm surprised they work... I mean [this](https://godbolt.org/z/vxs1rY9j7) fails with `--std=c++17` but I have just found that succeeds with `--std=c++20`, why?

Reply:
Pretty sure it's because C++20 introduced the rewriting rules that makes this work: you no longer need to write both variants (x,y) and (y,x) for == and other operators, just one is enough and the other one will be synthesized for you (it came along with all the spaceship stuff).

Seastar probably has a problem because they are still compiling for C++17?

Reply:
See also this earlier comment:

https://github.com/redpanda-data/redpanda/pull/10385#discussion_r1189065925